firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java

```diff
@@ -4578,22 +4578,324 @@ public void testGetWaitsForConnection() throws DatabaseException, InterruptedExc
+    long val = 34;
+    DatabaseReference node = db.getReference().push();
+
+    try {
+      await(node.setValue(val));
+      Semaphore semaphore = new Semaphore(0);
+      ValueEventListener listener =
+          new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot snapshot) {
+              assertEquals(val, snapshot.getValue());
+              semaphore.release();
+            }
+
+            @Override
+            public void onCancelled(@NonNull DatabaseError error) {}
+          };
+      node.addValueEventListener(listener);
+      IntegrationTestHelpers.waitFor(semaphore);
+      db.goOffline();
+      DataSnapshot snapshot = await(node.get());
+      assertEquals(val, snapshot.getValue());
+      node.removeEventListener(listener);
+      db.goOnline();
+    } catch (ExecutionException e) {
+      fail("get threw an exception: " + e);
+    }
+  }
+
+  @Test
+  public void testGetResolvesToCacheWhenOnlineAndChildLevelListener()
+      throws DatabaseException, InterruptedException {
+    FirebaseApp readApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseApp writeApp =
+        appForDatabaseUrl(IntegrationTestValues.getDatabaseUrl(), UUID.randomUUID().toString());
+    FirebaseDatabase readDb = FirebaseDatabase.getInstance(readApp);
+    FirebaseDatabase writeDb = FirebaseDatabase.getInstance(writeApp);
+    Semaphore semaphore = new Semaphore(0);
+    long val = 34;
+    String parentNodeKey = Objects.requireNonNull(writeDb.getReference().push().getKey());
+    DatabaseReference parentWriteNode = writeDb.getReference().child(parentNodeKey);
+    String childNodeKey = Objects.requireNonNull(parentWriteNode.push().getKey());
+    DatabaseReference childReadNode =
+        readDb.getReference().child(parentNodeKey).child(childNodeKey);
+    DatabaseReference childWriteNode = parentWriteNode.child(childNodeKey);
+    ValueEventListener listener =
+        new ValueEventListener() {
+          @Override
+          public void onDataChange(@NonNull DataSnapshot snapshot) {
+            semaphore.release();
+          }
+
+          @Override
+          public void onCancelled(@NonNull DatabaseError error) {
+            // no-op
+          }
+        };
+
+    try {
+      await(childWriteNode.setValue(val));
+      new ReadFuture(
+              childReadNode,
+              events -> {
+                assertEquals(1, events.size());
+                assertEquals(val, events.get(0).getSnapshot().getValue());
+                return true;
+              })
+          .timedGet();
+      childReadNode.addValueEventListener(listener);
+      DatabaseReference parentReadNode = readDb.getReference().child(parentNodeKey);
+      IntegrationTestHelpers.waitFor(semaphore);
+      DataSnapshot snapshot = await(parentReadNode.get());
+      assertEquals(snapshot.child(childNodeKey).getValue(), val);
```

Comment:
Just a note, no need to make any changes.

I think in the future for assertions like this we should assert the contents of `snapshot` as opposed to indexing into it with `.child()`:
```
assertEquals(MapBuilder.put(childNodeKey, val).build(), snapshot.getValue());
```
I think in general it's easier to think about "values of references" as on L4765 below. It's also more correct because it tests that unexpected keys don't exist at `snapshot.getValue()`.